### DISCUSSION TITLE
stdout on end

### DISCUSSION THREAD

**Discussion by plondon** [raxod502/riju#132] [General]
Hi @raxod502! I'm trying to fire an event after the stdout data has finished streaming. I think I need to hook into something here: https://github.com/raxod502/riju/blame/master/backend/api.js#L328

Do you know of a simple way to do that? Or do you have any advice? Thanks!

**Comment by raxod502** [comment]
You mean from the frontend? You should probably look for the `serviceFailed` event for the `terminal` service.

![image](https://user-images.githubusercontent.com/6559064/138534063-2cc2d1c1-543c-4b93-b8e9-f17921c8b2ba.png)

Or are you talking about modifying the backend? You could set a listener for the EOF event on the stdout stream, but that's approximately equivalent for most realistic use cases to just listening for process termination, which we do already.

**Reply by plondon** [reply]
Gotcha, yes this is what I'm looking for. However, this event only fires if the language does not have a repl. Any advice on handling languages w/ repls? Much appreciated in advance!

**Reply by raxod502** [reply]
Ah. You want to tell when the user code has finished executing. Unfortunately there's no real way to do that at least with the current architecture, since it's a single program execution. For example, if you run `python3 -i main.py`, there's no way to tell when `main.py` ends and when the repl begins.

Open to suggestions on how that could be changed, but I don't think it's possible without some rearchitecting.

**Reply by plondon** [reply]
I think it would be possible to use spawnSync instead of spawn in `runCode`, but an initial pass did not work. I can keep investigating. Glad to know there is no real way to do what I'm trying at the moment!

**Reply by raxod502** [reply]
If you use `spawnSync`, then all that's going to happen is the invocation will never terminate. This isn't something you can change at the level of process invocation: you'd have to fundamentally change the semantic interpretation of the `repl` and `run` commands, and rewrite them for all existing languages.
